From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pranav Vashi <neobuddy89@gmail.com>
Date: Fri, 2 Oct 2026 16:10:15 +0530
Subject: [PATCH 08/10] Launcher3: QuickstepTransitionManager: Fix and optimize
 transition animators

- Reuse crop rects in the app-launch and fallback closing animators
  instead of allocating a new Rect every frame, with null guards
- Keep one SurfaceTransaction per frame to avoid racing the
  RenderThread merge in SurfaceTransactionApplier
- Normalize rotationChange to [0, 3] so negative deltas still take
  the 90/270 degree paths
- Guard a null screenSpaceBounds in the fallback closing animator
- Always release LauncherBackAnimationController on destroy, and
  null-check it in onOverviewTargetChange()
- Fix indentation in getActivityLaunchOptions() and rename
  bankAnimState to backAnimState

[MikolajQ: Taken from crDroid 17.0 (803e928a14) in place of the Android 16 patches 0008 and 0011 (per-frame allocations in QuickstepTransitionManager)]

Co-authored-by: John Galt <johngaltfirstrun@gmail.com>
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 .../launcher3/QuickstepTransitionManager.java | 79 +++++++++++--------
 1 file changed, 48 insertions(+), 31 deletions(-)

diff --git a/quickstep/src/com/android/launcher3/QuickstepTransitionManager.java b/quickstep/src/com/android/launcher3/QuickstepTransitionManager.java
index 4fe7c2c..ca2ccc7 100644
--- a/quickstep/src/com/android/launcher3/QuickstepTransitionManager.java
+++ b/quickstep/src/com/android/launcher3/QuickstepTransitionManager.java
@@ -468,9 +468,9 @@ public class QuickstepTransitionManager implements OnDeviceProfileChangeListener
 
         long statusBarTransitionDelay = duration - STATUS_BAR_TRANSITION_DURATION
                 - STATUS_BAR_TRANSITION_PRE_DELAY;
-      ActivityOptions options = ActivityOptions.makeRemoteAnimation(
-              new RemoteAnimationAdapter(appLaunchRunner, duration, statusBarTransitionDelay),
-              remoteTransition);
+        ActivityOptions options = ActivityOptions.makeRemoteAnimation(
+                new RemoteAnimationAdapter(appLaunchRunner, duration, statusBarTransitionDelay),
+                remoteTransition);
         IRemoteCallback endCallback = completeRunnableListCallback(
                 onEndCallback, mLauncher, MAIN_EXECUTOR);
         options.setOnAnimationAbortListener(endCallback);
@@ -887,6 +887,7 @@ public class QuickstepTransitionManager implements OnDeviceProfileChangeListener
                 null /* fadeOutView */, !appTargetsAreTranslucent, launcherIconBounds,
                 true /* isOpening */);
         Rect crop = new Rect();
+        Rect closingTargetCrop = new Rect();
         Matrix matrix = new Matrix();
 
         SurfaceTransactionApplier surfaceApplier =
@@ -1109,20 +1110,24 @@ public class QuickstepTransitionManager implements OnDeviceProfileChangeListener
                         } else {
                             tmpPos.set(surface.position.x, surface.position.y);
                         }
-                        final Rect crop = new Rect(surface.screenSpaceBounds);
-                        crop.offsetTo(0, 0);
+                        if (surface.screenSpaceBounds != null) {
+                            closingTargetCrop.set(surface.screenSpaceBounds);
+                            closingTargetCrop.offsetTo(0, 0);
+                        } else {
+                            closingTargetCrop.setEmpty();
+                        }
 
-                        if ((rotationChange % 2) == 1) {
-                            int tmp = crop.right;
-                            crop.right = crop.bottom;
-                            crop.bottom = tmp;
+                        if ((rotationChange % 2) != 0) {
+                            int tmp = closingTargetCrop.right;
+                            closingTargetCrop.right = closingTargetCrop.bottom;
+                            closingTargetCrop.bottom = tmp;
                             tmp = tmpPos.x;
                             tmpPos.x = tmpPos.y;
                             tmpPos.y = tmp;
                         }
                         matrix.setTranslate(tmpPos.x, tmpPos.y);
                         builder.setMatrix(matrix)
-                                .setWindowCrop(crop)
+                                .setWindowCrop(closingTargetCrop)
                                 .setAlpha(1f);
                     }
                 }
@@ -1543,6 +1548,9 @@ public class QuickstepTransitionManager implements OnDeviceProfileChangeListener
      * Called when the overview-target changes. Updates the back callback registration state.
      */
     public void onOverviewTargetChange() {
+        if (mBackAnimationController == null) {
+            return;
+        }
         if (isHomeRoleHeld()) {
             mBackAnimationController.registerBackCallbacks(mHandler);
         } else {
@@ -1570,18 +1578,16 @@ public class QuickstepTransitionManager implements OnDeviceProfileChangeListener
 
     protected void unregisterRemoteTransitions() {
         SystemUiProxy.INSTANCE.get(mLauncher).unshareTransactionQueue();
-        if (SEPARATE_RECENTS_ACTIVITY.get()) {
-            return;
-        }
-        if (mLauncherOpenTransition == null) return;
-        SystemUiProxy.INSTANCE.get(mLauncher).unregisterRemoteTransition(
-                mLauncherOpenTransition);
-        mLauncherOpenTransition = null;
-        mWallpaperOpenTransitionRunner = null;
-        if (mMoveDisplayTransition != null) {
-            SystemUiProxy.INSTANCE.get(mLauncher)
-                    .unregisterRemoteTransition(mMoveDisplayTransition);
-            mMoveDisplayTransition = null;
+        if (!SEPARATE_RECENTS_ACTIVITY.get() && mLauncherOpenTransition != null) {
+            SystemUiProxy.INSTANCE.get(mLauncher).unregisterRemoteTransition(
+                    mLauncherOpenTransition);
+            mLauncherOpenTransition = null;
+            mWallpaperOpenTransitionRunner = null;
+            if (mMoveDisplayTransition != null) {
+                SystemUiProxy.INSTANCE.get(mLauncher)
+                        .unregisterRemoteTransition(mMoveDisplayTransition);
+                mMoveDisplayTransition = null;
+            }
         }
         if (mBackAnimationController != null) {
             mBackAnimationController.unregisterBackCallbacks();
@@ -1625,7 +1631,7 @@ public class QuickstepTransitionManager implements OnDeviceProfileChangeListener
                 rotationChange = surface.rotationChange;
             }
         }
-        return rotationChange;
+        return ((rotationChange % 4) + 4) % 4;
     }
 
     /**
@@ -1856,6 +1862,7 @@ public class QuickstepTransitionManager implements OnDeviceProfileChangeListener
         Matrix matrix = new Matrix();
         Point tmpPos = new Point();
         Rect tmpRect = new Rect();
+        Rect fallbackCrop = new Rect();
         ValueAnimator closingAnimator = ValueAnimator.ofFloat(0, 1);
         int duration = CLOSING_TRANSITION_DURATION_MS;
         float windowCornerRadius = getWindowCornerRadius(mLauncher);
@@ -1886,14 +1893,24 @@ public class QuickstepTransitionManager implements OnDeviceProfileChangeListener
                         tmpPos.set(target.position.x, target.position.y);
                     }
 
-                    final Rect crop = new Rect(target.localBounds);
-                    crop.offsetTo(0, 0);
+                    if (target.localBounds != null) {
+                        fallbackCrop.set(target.localBounds);
+                        fallbackCrop.offsetTo(0, 0);
+                    } else {
+                        fallbackCrop.setEmpty();
+                    }
                     if (target.mode == MODE_CLOSING) {
-                        tmpRect.set(target.screenSpaceBounds);
+                        if (target.screenSpaceBounds != null) {
+                            tmpRect.set(target.screenSpaceBounds);
+                        } else {
+                            tmpRect.set(tmpPos.x, tmpPos.y,
+                                    tmpPos.x + fallbackCrop.width(),
+                                    tmpPos.y + fallbackCrop.height());
+                        }
                         if ((rotationChange % 2) != 0) {
-                            final int right = crop.right;
-                            crop.right = crop.bottom;
-                            crop.bottom = right;
+                            final int right = fallbackCrop.right;
+                            fallbackCrop.right = fallbackCrop.bottom;
+                            fallbackCrop.bottom = right;
                         }
                         matrix.setScale(mScale.value, mScale.value,
                                 tmpRect.centerX(),
@@ -1901,14 +1918,14 @@ public class QuickstepTransitionManager implements OnDeviceProfileChangeListener
                         matrix.postTranslate(0, mDy.value);
                         matrix.postTranslate(tmpPos.x, tmpPos.y);
                         builder.setMatrix(matrix)
-                                .setWindowCrop(crop)
+                                .setWindowCrop(fallbackCrop)
                                 .setAlpha(mAlpha.value)
                                 .setCornerRadius(windowCornerRadius)
                                 .setShadowRadius(mShadowRadius.value);
                     } else if (target.mode == MODE_OPENING) {
                         matrix.setTranslate(tmpPos.x, tmpPos.y);
                         builder.setMatrix(matrix)
-                                .setWindowCrop(crop)
+                                .setWindowCrop(fallbackCrop)
                                 .setAlpha(1f);
                     }
                 }
